# CMI dataset

Detecting behavior from sensor data for kaggle competition

 Background Knowledge , for better understanding of the problem statement and the data.

 

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

In [3]:
train_demo_df=pd.read_csv(r"data/train_demographics.csv")
train_df=pd.read_csv(r"data/train.csv")

In [18]:
test_df_demo_df=pd.read_csv(r"data/test_demographics.csv")
test_df=pd.read_csv(r"data/test.csv")

In [4]:
train_demo_df

,subject,adult_child,age,sex,handedness,height_cm,shoulder_to_wrist_cm,elbow_to_wrist_cm
0,SUBJ_000206,1,41,1,1,172.0,50,25.0
1,SUBJ_001430,0,11,0,1,167.0,51,27.0
2,SUBJ_002923,1,28,1,0,164.0,54,26.0
3,SUBJ_003328,1,33,1,1,171.0,52,25.0
4,SUBJ_004117,0,15,0,1,184.0,54,28.0
...,...,...,...,...,...,...,...,...
76,SUBJ_063319,1,25,1,1,173.0,50,29.0
77,SUBJ_063346,1,42,1,1,183.0,51,26.0
78,SUBJ_063447,1,36,0,1,175.0,54,25.0
79,SUBJ_063464,0,15,0,1,162.0,43,23.0


In [21]:
[i for i in train_df.columns if i not in test_df.columns]

['sequence_type', 'orientation', 'behavior', 'phase', 'gesture']

In [23]:
train_df.groupby(["sequence_id","orientation","gesture","sequence_type"]).count()

,,,,row_id,sequence_counter,subject,behavior,phase,acc_x,acc_y,acc_z,rot_w,rot_x,...,tof_5_v54,tof_5_v55,tof_5_v56,tof_5_v57,tof_5_v58,tof_5_v59,tof_5_v60,tof_5_v61,tof_5_v62,tof_5_v63
sequence_id,orientation,gesture,sequence_type,,,,,,,,,,,,,,,,,,,,,
SEQ_000007,Seated Lean Non Dom - FACE DOWN,Cheek - pinch skin,Target,57,57,57,57,57,57,57,57,57,57,...,57,57,57,57,57,57,57,57,57,57
SEQ_000008,Seated Lean Non Dom - FACE DOWN,Forehead - pull hairline,Target,68,68,68,68,68,68,68,68,68,68,...,68,68,68,68,68,68,68,68,68,68
SEQ_000013,Lie on Side - Non Dominant,Cheek - pinch skin,Target,53,53,53,53,53,53,53,53,53,53,...,53,53,53,53,53,53,53,53,53,53
SEQ_000016,Seated Lean Non Dom - FACE DOWN,Write name on leg,Non-Target,61,61,61,61,61,61,61,61,61,61,...,61,61,61,61,61,61,61,61,61,61
SEQ_000018,Seated Lean Non Dom - FACE DOWN,Forehead - pull hairline,Target,54,54,54,54,54,54,54,54,54,54,...,54,54,54,54,54,54,54,54,54,54
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
SEQ_065508,Lie on Back,Text on phone,Non-Target,76,76,76,76,76,76,76,76,76,76,...,76,76,76,76,76,76,76,76,76,76
SEQ_065519,Seated Straight,Neck - scratch,Target,84,84,84,84,84,84,84,84,84,84,...,84,84,84,84,84,84,84,84,84,84
SEQ_065522,Lie on Back,Above ear - pull hair,Target,78,78,78,78,78,78,78,78,78,78,...,78,78,78,78,78,78,78,78,78,78


The columns in train gesture,orientation, sequence_id are basically repeated for each sequence given in the data and cross checked from the numbers of columns when grouped by.

But the columns that are to be predicted are behavior , phase

In [13]:
train_df.columns

Index(['row_id', 'sequence_type', 'sequence_id', 'sequence_counter', 'subject',
       'orientation', 'behavior', 'phase', 'gesture', 'acc_x',
       ...
       'tof_5_v54', 'tof_5_v55', 'tof_5_v56', 'tof_5_v57', 'tof_5_v58',
       'tof_5_v59', 'tof_5_v60', 'tof_5_v61', 'tof_5_v62', 'tof_5_v63'],
      dtype='str', length=341)

In [ ]:
train_df.sequence_id.value_counts()
'''
 we have got varying sequence lengths, two approaches as of now:
     1. Padding sequences to maximum length
    2. Using RNNs that can handle variable length sequences (like LSTM, GRU).
    3.7
 '''

sequence_id
SEQ_015261    700
SEQ_014205    671
SEQ_028188    647
SEQ_060593    630
SEQ_016031    567
             ... 
SEQ_011271     34
SEQ_022667     34
SEQ_032114     34
SEQ_059162     34
SEQ_009199     29
Name: count, Length: 8151, dtype: int64

Time series data is presented to us, each instance of sequences is given.

The dataset has 8151 sequences and 81 samples or test subjects.

This problem has to be solved to using Recurrent Neural networks, as the data is sequential and carries information across time steps. Have to look closely to find more about the data and the problem statement.